transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionMetadata.java

```diff
@@ -0,0 +1,732 @@
+        this.prevProducerId = prevProducerId;
+        this.nextProducerId = nextProducerId;
+        this.producerEpoch = producerEpoch;
+        this.lastProducerEpoch = lastProducerEpoch;
+        this.txnTimeoutMs = txnTimeoutMs;
+        this.state = state;
+        this.topicPartitions = new HashSet<>(topicPartitions);
+        this.txnStartTimestamp = txnStartTimestamp;
+        this.txnLastUpdateTimestamp = txnLastUpdateTimestamp;
+        this.clientTransactionVersion = clientTransactionVersion;
+        this.pendingState = Optional.empty();
+        this.hasFailedEpochFence = false;
+        this.lock = new ReentrantLock();
+    }
+
+    public <T> T inLock(Supplier<T> function) {
+        lock.lock();
+        try {
+            return function.get();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void addPartitions(Collection<TopicPartition> partitions) {
+        topicPartitions.addAll(partitions);
+    }
+
+    public void removePartition(TopicPartition topicPartition) {
+        if (state != TransactionState.PREPARE_COMMIT && state != TransactionState.PREPARE_ABORT)
+            throw new IllegalStateException("Transaction metadata's current state is " + state + ", and its pending state is " +
+                pendingState + " while trying to remove partitions whose txn marker has been sent, this is not expected");
+
+        topicPartitions.remove(topicPartition);
+    }
+
+    // this is visible for test only
+    public TxnTransitMetadata prepareNoTransit() {
+        // do not call transitTo as it will set the pending state, a follow-up call to abort the transaction will set its pending state
+        return new TxnTransitMetadata(producerId, prevProducerId, nextProducerId, producerEpoch, lastProducerEpoch, txnTimeoutMs,
+            state, Set.copyOf(topicPartitions), txnStartTimestamp, txnLastUpdateTimestamp, clientTransactionVersion);
+    }
+
+    public TxnTransitMetadata prepareFenceProducerEpoch() {
+        if (producerEpoch == Short.MAX_VALUE)
+            throw new IllegalStateException("Cannot fence producer with epoch equal to Short.MaxValue since this would overflow");
+
+        // If we've already failed to fence an epoch (because the write to the log failed), we don't increase it again.
+        // This is safe because we never return the epoch to client if we fail to fence the epoch
+        short bumpedEpoch = hasFailedEpochFence ? producerEpoch : (short) (producerEpoch + 1);
+
+        return prepareTransitionTo(
+            TransactionState.PREPARE_EPOCH_FENCE,
+            producerId,
+            nextProducerId,
+            bumpedEpoch,
+            lastProducerEpoch,
+            txnTimeoutMs,
+            Set.copyOf(topicPartitions),
+            txnStartTimestamp,
+            txnLastUpdateTimestamp,
+            clientTransactionVersion
+        );
+    }
+
+    public Either<Errors, TxnTransitMetadata> prepareIncrementProducerEpoch(
```

Comment:
it can throw exception and callers can convert the exception to error. we don't need to introduce Either I think